### ISSUE TITLE
Wrong IsLatest value due to bad DB connection during draft saving in Elsa 2

### ISSUE THREAD

**Issue by alessandro-gandolfi-ors** [elsa-workflows/elsa-core#5146]
Since the function to save a new draft doesn't use transactions (as far as I understood, I could be wrong), sometimes it can produce bad situations on the DB, where the "latest" workflow definition has already been changed with IsLatest = 0 but the new draft hasn't been created due to an exception (for example, a bad connection and/or timeout).

The version of Elsa 2 I'm using is 2.11.0 but I have checked and no other changes have been made to the files involved since that version to 2.14.1.

After this error, that definition can't be modified anymore, since there are no definitions with IsLatest = 1 and Elsa can't figure out the new version that should be used, trying to use 1 as new version and not respecting the index IX_WorkflowDefinition_DefinitionId_VersionId on the DB (version 1 is already present).

The first few lines of the exception are usually like this, I just replaced the DefinitionId not to make confusion:

```
 An unhandled exception has occurred while executing the request.
       Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while saving the entity changes. See the inner exception for details.
        ---> Microsoft.Data.SqlClient.SqlException (0x80131904): Cannot insert duplicate key row in object 'Elsa.WorkflowDefinitions' with unique index 'IX_WorkflowDefinition_DefinitionId_VersionId'. The duplicate key value is (XXXXXXXXXXXXXXXXXXXXXXX, 1).
       The statement has been terminated.
```

This has happened multiple times, sometimes for bad connection to the DB and other times for reason I couldn't understand from the logs, even when there were no errors regarding the connection.

![Screenshot 2024-03-28 110347](https://github.com/elsa-workflows/elsa-core/assets/95684590/e2f28a86-e523-435d-b319-b04c72cc7a5d)
![image](https://github.com/elsa-workflows/elsa-core/assets/95684590/86a73eb3-0b38-4e45-bca3-190d8fe325fe)

In the last image there is an example of the versions of a definition where the published version has been modified, so the BE tried and succeded in updating the IsLatest flag to false, but then the new draft hasn't been created.

I have seen that a PR has been merged (#3752) to invert the order of the operations for Elsa 3, backporting the changes to Elsa 2, if possible, could help a lot, even without transactions.

So, if I understood correctly, these operations should be atomic but transactions have not been used. Is there a reason behind this choice or have I the wrong idea about it?

Thanks in advance.

**Comment by sfmskywalker** [comment]
Elsa 3 has mastered this logic flawlessly. You might consider backporting the logic from there and submitting a pull request.

**Comment by alessandro-gandolfi-ors** [comment]
I'm asking my boss if I can spend time working on it, it would help us but there is already the idea to migrate to elsa 3 in the future...

Meanwhile, could you elaborate a bit on why those operations are not atomic?

**Comment by sfmskywalker** [comment]
Absolutely, the abstraction of the persistence layer—encompassing EF Core, MongoDB, Dapper, and possible third-party implementations—makes handling this from the application layer quite practical. Given the smooth operation with Elsa 3 and workflow publications, there hasn't been a pressing need to alter this method. Nevertheless, I'm very interested in any suggestions for integrating this logic within a transaction that's compatible across all persistence providers. Your insights on achieving this would be greatly appreciated.

**Comment by sfmskywalker** [comment]
Closing because this issue targets Elsa 2 or the legacy dashboard stack, which is no longer an actively supported target in this repository. Current UI work happens in Elsa Studio for Elsa 3. If you can still reproduce the underlying problem on a supported Elsa 3 + Elsa Studio setup, please open a new issue with fresh reproduction steps.
